File: src/xian_aqc_utils/src/xian_spreader_images_show.py
import rospy
import cv2
import numpy as np
import datetime
from cv_bridge import CvBridge
from sensor_msgs.msg import Image
from xian_msg_pkg.msg import xian_spreader_images_msg

class XianSpreaderImagesShow:
    def __init__(self):
        self.bridge = CvBridge()
        self.counter = 0

        # Initialize ROS node
        rospy.init_node('xian_spreader_images_show_node', anonymous=True)

        # Subscribe to the topic
        self.command_subscribe = rospy.Subscriber('xian_aqc_spreader_images', xian_spreader_images_msg, self.command_callback)

        # Create a wall timer for heartbeat
        self.m_timer_HeartBeat = rospy.Timer(rospy.Duration(1.0), self.m_timer_HeartBeat_f)

        # Initialize time variables
        self.cur_time = rospy.Time.now()
        self.pre_time = rospy.Time.now()
        self.timediff = 1.0

        # Initialize OpenCV windows
        # cv2.namedWindow("xian_spreader_images_show01", cv2.WINDOW_NORMAL)

    def m_timer_HeartBeat_f(self, event):
        self.counter = self.counter + 1 if self.counter < 1000 else 0
        rospy.loginfo(f"xian_spreader_images_show_heart_beat: {self.counter}")

    def command_callback(self, data):
        timeStr = datetime.datetime.now().strftime("%Y%m%d%H%M%S")
        self.pre_time = self.cur_time
        self.cur_time = rospy.Time.now()

        try:
            tl_image = self.bridge.imgmsg_to_cv2(data.tl_image, "bgr8")
            tr_image = self.bridge.imgmsg_to_cv2(data.tr_image, "bgr8")
            bl_image = self.bridge.imgmsg_to_cv2(data.bl_image, "bgr8")
            br_image = self.bridge.imgmsg_to_cv2(data.br_image, "bgr8")

            # Merging images
            mask_merge_col_0 = np.vstack((tl_image, bl_image))
            mask_merge_col_1 = np.vstack((tr_image, br_image))
            merge_row1 = np.hstack((mask_merge_col_0, mask_merge_col_1))

            # Resize using CUDA
            merge_row1_gpu = cv2.cuda_GpuMat()
            merge_row1_resize_gpu = cv2.cuda_GpuMat()
            merge_row1_gpu.upload(merge_row1)
            merge_row1_resize_gpu = cv2.cuda.resize(merge_row1_gpu, (merge_row1.shape[1] // 4, merge_row1.shape[0] // 4))
            merge_row1_resize = merge_row1_resize_gpu.download()
            
            # merge_row1_resize = cv2.resize(merge_row1, (merge_row1.shape[1] // 4, merge_row1.shape[0] // 4))

            if merge_row1_resize is not None:
                cv2.imshow("xian_spreader_images_show01", merge_row1_resize)
                cv2.waitKey(10)
                print(merge_row1_resize.shape)
                pass
            else:
                rospy.logerr("Failed to load image")

            # Calculate FPS
            elapsedTimeP = (self.cur_time - self.pre_time).to_sec() * 1000
            self.timediff = elapsedTimeP
            rospy.loginfo(f"FPS: {1000.0 / self.timediff}")

        except Exception as e:
            rospy.logerr(f"Error in processing images: {e}")

if __name__ == '__main__':
    try:
        xian_spreader_images_show = XianSpreaderImagesShow()
        rospy.spin()
    except rospy.ROSInterruptException:
        pass
    finally:
        cv2.destroyAllWindows()

File: src/xian_aqc_utils/src/xian_spreader_images_show.cpp
#include<ros/ros.h>
#include<cv_bridge/cv_bridge.h>
#include<sensor_msgs/image_encodings.h>
#include<image_transport/image_transport.h>

#include<opencv2/opencv.hpp>
#include<opencv2/highgui/highgui.hpp>
#include<opencv2/imgproc/imgproc.hpp>

// #include <opencv2/cudaarithm.hpp>
// #include <opencv2/cudaimgproc.hpp>
// #include <opencv2/cudawarping.hpp>

#include<stdio.h>
#include<sys/types.h>
#include "xian_msg_pkg/xian_spreader_images_msg.h"
#include "zpmc_cv_control.h"


class Xian_SpreaderImagesShow
{
    public:
        Xian_SpreaderImagesShow()
        {
            // 创建一个ROS节点句柄
            ros::NodeHandle nh;
            command_subscribe_ = nh.subscribe<xian_msg_pkg::xian_spreader_images_msg>("xian_aqc_spreader_images", 1, &Xian_SpreaderImagesShow::command_callback, this);

        }

        ros::WallTimer m_timer_HeartBeat;

        void m_timer_HeartBeat_f(const ros::WallTimerEvent& event)
        {
            counter = counter > 1000 ? 0 : (counter + 1);
            std::cout << "xian_spreader_images_show_heart_beat: " << counter << std::endl;
        }

    private:
        ros::Subscriber command_subscribe_;

        std::chrono::_V2::system_clock::time_point cur_time = std::chrono::high_resolution_clock::now();
        std::chrono::_V2::system_clock::time_point pre_time = std::chrono::high_resolution_clock::now();
        std::chrono::__enable_if_is_duration<std::chrono::duration<long int, std::ratio<1, 1000>>> elapsedTimeP;
        double timediff = 1.0;
        int counter = 0;
        std::string timeStr;
 
        cv::Mat tl_image, tr_image, bl_image, br_image, merge_row1_resize;
        sensor_msgs::Image spreder_imgs_resize;

        void command_callback(const xian_msg_pkg::xian_spreader_images_msgConstPtr& data)
        {
            timeStr = zpmc::zpmc_get_stystem_time();
            pre_time = cur_time;
            cur_time = std::chrono::high_resolution_clock::now();

            tl_image = cv_bridge::toCvShare(data->tl_image, data, "bgr8")->image; 
            tr_image = cv_bridge::toCvShare(data->tr_image, data, "bgr8")->image; 
            bl_image = cv_bridge::toCvShare(data->bl_image, data, "bgr8")->image;
            br_image = cv_bridge::toCvShare(data->br_image, data, "bgr8")->image;

            cv::Mat mask_merge_col_0 = zpmc::zpmc_images_merge_row(tl_image, bl_image);
            cv::Mat mask_merge_col_1 = zpmc::zpmc_images_merge_row(tr_image, br_image);
            
            cv::Mat merge_row1 = zpmc::zpmc_images_merge_col(mask_merge_col_0, mask_merge_col_1);

            // cv::cuda::GpuMat merge_row1_gpu, merge_row1_resize_gpu;
            // merge_row1_gpu.upload(merge_row1);
            // cv::cuda::resize(merge_row1_gpu, merge_row1_resize_gpu, cv::Size((int)(merge_row1.cols/4), (int)(merge_row1.rows/4)), 2); 
            // merge_row1_resize_gpu.download(merge_row1_resize);
            cv::resize(merge_row1, merge_row1_resize, cv::Size((int)(merge_row1.cols/4), (int)(merge_row1.rows/4)), 2);
            if (merge_row1_resize.empty()) 
            {
                std::cerr << "Failed to load image" << std::endl;
            }
            else
            {

                cv::imshow("xian_spreader_images_show01:", merge_row1_resize);
                cv::waitKey(10);


                //cv::imwrite("/root/code/xian_aqc_ws/xian_project_file/trt/results/"+timeStr+"_tl_.jpg", tl_image);
                //cv::imwrite("/root/code/xian_aqc_ws/xian_project_file/trt/results/"+timeStr+"tr_.jpg", tr_image);
                //cv::imwrite("/root/code/xian_aqc_ws/xian_project_file/trt/results/"+timeStr+"bl_.jpg", bl_image);
                //cv::imwrite("/root/code/xian_aqc_ws/xian_project_file/trt/results/"+timeStr+"br_.jpg", br_image);
                // std::cerr << "iiiiiiiiiiiiiiiiiiiiii" << std::endl;
            }
            

            elapsedTimeP = std::chrono::duration_cast<std::chrono::milliseconds>(cur_time - pre_time);
            timediff = elapsedTimeP.count();
            std::cout << "FPS: " << 1000.0 / timediff << std::endl;
        } 


        

        
            
};

int main(int argc, char** argv)
{
    //initial and name node
    ros::init(argc,argv,"xian_spreader_images_show_node");
    Xian_SpreaderImagesShow xian_spreader_images_show;

    // 创建一个ROS节点句柄
    ros::NodeHandle nh_2;
    ros::AsyncSpinner spinner(0);
    spinner.start();

    xian_spreader_images_show.m_timer_HeartBeat = nh_2.createWallTimer(ros::WallDuration(1.0), &Xian_SpreaderImagesShow::m_timer_HeartBeat_f, &xian_spreader_images_show);

    ros::waitForShutdown();
    
    // ros::spin();
    return 0;
}
